src/kepler/ipc.zig

```diff
@@ -0,0 +1,582 @@
+            return;
+        }
+        // At this point, all non-owning references are lost
+        // If death note was already sent, we can decrement owning count
+        // as no more client references will be sent
+        if (@atomicLoad(bool, &self.death_note_sent, .Acquire)) {
+            self.decrement_internal();
+            return;
+        }
+        // If it wasn't set, we don't want to decrement owning-refcount,
+        // as note will still store client (non-ownning) references to the object
+        @atomicStore(bool, &self.death_note_sent, true, .Release);
+        // Check if server is already dead. If it die later,
+        // server notification queue will clean everything up anyways
+        if (@atomicLoad(bool, &self.dying, .Acquire)) {
+            // If we can't send a reference, just terminate
+            self.decrement_internal();
+            return;
+        }
+        // Send ping of death note
+        self.death_note.typ = .EndpointUnreachable;
+        self.death_note.owner_ref = .{ .endpoint = self.borrow() };
+        // If send failed, we can just ignore the failure
+        // drop() called in send() will see that message was already sent
+        // and will terminate
+        self.queue.send(&self.death_note) catch {};
+    }
+
+    /// Drop owning reference to the endpoint
+    pub fn shutdown(self: *@This()) void {
+        // Indicate that we shut down the server
+        @atomicStore(bool, &self.dying, true, .Release);
+        // Decrement internal reference count
+        self.decrement_internal();
+    }
+
+    /// Decrement internal owning reference count
+    fn decrement_internal(self: *@This()) void {
+        if (@atomicRmw(usize, &self.own_ref_count, .Sub, 1, .AcqRel) > 1) {
+            return;
+        }
+        self.deinit();
+    }
+
+    /// Dispose endpoint object
+    fn deinit(self: *@This()) void {
+        self.queue.drop();
+        self.allocator.destroy(self);
+    }
+
+    /// Return true if endpoint still listens for incoming connections
+    fn is_active(self: *const @This()) bool {
+        return !@atomicLoad(bool, &self.dying, .Acquire);
+    }
+};
+
+/// Stream is an object that allows two peers to exchange
+/// notification with each other. Each stream has only two peers -
+/// .Producer and .Consumer
+pub const Stream = struct {
+    /// Peer type
+    pub const Peer = enum(u1) {
+        /// Consumer is the peer that initially sent a request
+        /// to make a stream
+        Consumer = 0,
+        /// Producer is the peer that accepted request to make
+        /// a stream
+        Producer = 1,
+
+        // Convert to index
+        pub fn idx(self: Peer) u1 {
+            return @enumToInt(self);
+        }
+
+        // Get enum member for other peer
+        pub fn other(self: Peer) Peer {
+            return @intToEnum(Peer, 1 - self.idx());
+        }
+    };
+
+    /// Peer connection status
+    pub const PeerStatus = enum(usize) {
```

Comment:
What is the reason for `enum(usize)` over `enum` here?

Reply:
https://github.com/ziglang/zig/issues/7976

Reply:
Nice, please reference the issue from the code so that we know when it's fixed and can update the code.